arkouda/join.py

```diff
@@ -205,6 +210,29 @@ def inner_join(
     """
     from inspect import signature
 
+    if type(left) != type(right):
+        raise TypeError("Left and Right arrays must be of same type")
+
+    # Convert Strings and Categorical left & right values to Categorical.codes. This generates an
+    # index representation of the strings. This way they can be handled as numerics and follow the
+    # same process as int
+    t = cast(str, pdarray.objtype)
+    cats = None
+
+    if isinstance(left, Strings) and isinstance(right, Strings):
+        left = Categorical(left)
+        right = Categorical(right)
```

Comment:
This will result in more than double the data within strings being generated. With something that is already memory bound this is a bit concerning.

Reply:
This is something I'm aware of and intend to handle through #2390 by moving a lot of the setup workflow into Chapel where I can more easily operate on the `uint8` versions of the segString. This should improve performance a lot, but will take a lot longer to implement

Reply:
After touching base with @Ethan-DeBandi99 we decided to convert this PR to a draft for now and I'll go through the code with the intent to improve performance as much as possible instead of the initial plan of focusing on function here and performance in #2390. Once complete we'll look at the performance numbers again and close #2390 with this PR as well if we're happy with where they are